### PR TITLE
Confluence testing

### LINKED ISSUE #34: Run tests with user-defined rules
- [ ] Add a step to run tests with user-defined rules

### PR DESCRIPTION
**Pull request by aabounegm**
- Added confluence testing using QuickCheck and Hspec (with manual inputs)
- Fixed `applyRules` to only return unique results, regardless of the order of attributes in the objects.
- Added an extra rule for application on `⊥`
- Added support for any object (not just `MetaId`s) for the `nf` condition
- Added support for the `current_attribute` context (useful for rules 1 and 2)
- Fixed the counting of `ν` attributes (by @deemp)
- Customized the `Show` instance of `Subst` to use `printTree`
- Fixed rule application in unit tests

Closes #34, #35, #5

---

## PR-Codex overview
This PR introduces QuickCheck, updates test titles, adds new rules and tests, modifies normalization functions, and enhances rule application logic.

### Detailed summary
- Added QuickCheck to dependencies
- Updated test titles in `eo-phi-normalizer/test/eo/phi/confluence.yaml`
- Added a new rule and tests in `eo-phi-normalizer/test/eo/phi/rules/yegor.yaml`
- Modified normalization functions in `eo-phi-normalizer/src/Language/EO/Phi/Normalize.hs`
- Enhanced rule application logic in various files

> The following files were skipped due to too many changes: `eo-phi-normalizer/src/Language/EO/Phi/Rules/Common.hs`, `eo-phi-normalizer/test/Language/EO/Rules/PhiPaperSpec.hs`, `eo-phi-normalizer/src/Language/EO/Phi/Rules/Yaml.hs`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: eo-phi-normalizer/test/confluence.hs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import Test.QuickCheck
+
+instance Arbitrary Attribute where
+  arbitrary =
+    oneof
+      [ pure Phi
+      , pure Rho
+      , pure Sigma
+      , pure VTX
+      , Label . LabelId <$> listOf (elements ['a' .. 'z'])
```
Comment (fizruk) [inline eo-phi-normalizer/test/confluence.hs:17]: I think empty lists should not be allowed here.

File: eo-phi-normalizer/test/confluence.hs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      ]
+
+instance Arbitrary Binding where
+  arbitrary = sized $ \n -> do
```
Comment (fizruk) [inline eo-phi-normalizer/test/confluence.hs:21]: I don't think you need `sized` here.

File: eo-phi-normalizer/src/Language/EO/Phi/Rules/Common.hs [hunk 3]
```diff
@@ -146,8 +146,8 @@ nuCount obj = count isNu (objectBindings obj) + sum (map (sum . map nuCount . va
   values (AlphaBinding _ obj') = [obj']
   values _ = []
 
-intToBytesObject :: Int -> Object
-intToBytesObject n = Formation [DeltaBinding $ Bytes $ insertDashes $ pad $ showHex n ""]
+intToBytes :: Int -> Bytes
+intToBytes n = Bytes $ insertDashes $ pad $ showHex n ""
```
Comment (fizruk) [inline eo-phi-normalizer/src/Language/EO/Phi/Rules/Common.hs:150]: This does not work for negative integers.

Comment (aabounegm) [reply eo-phi-normalizer/src/Language/EO/Phi/Rules/Common.hs:150]: Does it need to? It's only used with nonnegative numbers

Comment (aabounegm) [reply eo-phi-normalizer/src/Language/EO/Phi/Rules/Common.hs:150]: Ok, just realized that an `arbitrary` integer can be negative. Will fix that

**Pull request merged**